src/ReinforcementLearningZoo/src/algorithms/offline_rl/NFQ.jl

```diff
@@ -0,0 +1,78 @@
+"""
+    NFQ{A<:AbstractApproximator, F, R} <: AbstractLearner
+    NFQ(approximator::A, num_iterations::Integer epochs::Integer, loss_function::F, batch_size::Integer, rng::R, γ::Float32) where {A<:AbstractApproximator, F, R}
+Neural Fitted Q-iteration as implemented in [1]
+
+# Keyword arguments
+- `approximator::AbstractApproximator` neural network
+- `num_iterations::Integer` number of value iteration iterations in FQI loop (i.e. the outer loop)
+- `epochs` number of epochs to train neural network per iteration
+- `loss_function::F` loss function of the NN
+- `sampler::BatchSampler{SARTS}` data sampler
+- `rng::R` random number generator
+- `γ::Float32` discount rate
+
+# References
+[1] Riedmiller, M. (2005). Neural Fitted Q Iteration – First Experiences with a Data Efficient Neural Reinforcement Learning Method. In: Gama, J., Camacho, R., Brazdil, P.B., Jorge, A.M., Torgo, L. (eds) Machine Learning: ECML 2005. ECML 2005. Lecture Notes in Computer Science(), vol 3720. Springer, Berlin, Heidelberg. https://doi.org/10.1007/11564096_32
+"""
+Base.@kwdef struct NFQ{A<:NeuralNetworkApproximator, F, R} <: AbstractLearner
+    approximator::A
+    num_iterations::Integer = 20
+    epochs::Integer = 100
+    loss_function::F = mse
+    rng::R = Random.GLOBAL_RNG
+    γ::Float32 = 0.9f0
+end
+
+function NFQ(;
+    approximator::A,
+    num_iterations::Integer = 20,
+    epochs::Integer = 1000,
+    loss_function::F = mse,
+    rng=Random.GLOBAL_RNG,
+    γ::Float32 = 0.9f0,
+    ) where {A<:NeuralNetworkApproximator, F}
+    NFQ(approximator, num_iterations, epochs, loss_function, rng, γ)
+end
+
+# Copied from BasicDQN but sure whether it's appropriate
+Flux.functor(x::NFQ) = (Q = x.approximator,), y -> begin
+    x = @set x.approximator = y.Q
+    x
+end
+
+function RLBase.plan!(learner::NFQ, env::AbstractEnv)
+    as = action_space(env)
+    return vcat(repeat(state(env), inner=(1, length(as))), transpose(as)) |> x -> send_to_device(device(learner.approximator), x) |> learner.approximator |> send_to_host |> vec
+end
+
+# Avoid optimisation in the middle of an episode
+function RLBase.optimise!(::NFQ, ::NamedTuple) end
```

Comment:
Instead of doing 'type piracy', I think we should discuss whether the optimize step belongs after the `::PostEpisodeStage`, I actually stumbled upon this 'quirk' today myself. What do you think @HenriDeh

Reply:
Why is it type piracy? The NFQ type and this function are defined in the Zoo package. 

However, I also think that the optimise is only called at the pre-act stage is a problem for ther many algorithms that optimise at the post-episode stage. We could extend the current implementation to `optimise!(usual_args..., ::AbstractStage)` and call it at both pre-act and post-episode in the run loop?

Reply:
I would also prefer to add the Stage to the optimise function and predefine optimse!(usual_args..., ::PostActStage) = nothing as a lot of algorithms train on whole episodes. I also thought to add a whole episode sampler for the trajectories.  (-> new issue for that?)

Reply:
>  I also thought to add a whole episode sampler.

Yes, I had the same thought, there's no way of doing that at the moment. 

Regarding optimise, however, I figured that we do not need to dispatch it on stages: it suffices to define the optimise! function like:
```
if is_terminated(env)
    #optimise etc.
end
```
and voila. Optimise is called right after `act!` so if the episode is over, we optimise, otherwise we go on.

Reply:
@HenriDeh Type piracy was clearly the wrong term, good point. I think we should resolve this 'issue' before addressing this pull request, here's a PR where we can focus on the `optimise` behavior...I think the trajectory `push` logic may need to be tweaked as well...https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/pull/899

Reply:
In my opinion, an algorithm might profit from dispatching on different stages:

A message exchange policy for example may exchange messages in order to optimise agent's policies and after the episode they look on the whole episode for example. Also, if the Stage is included in the optimise! call, the compiler might profit from this information and compile to a narrower level than using if-else switches, right?